presto-main/src/main/java/io/prestosql/server/protocol/Query.java

```diff
@@ -105,6 +106,7 @@
 
     private final QueryManager queryManager;
     private final QueryId queryId;
+    private final String slug = format("%016x%016x", ThreadLocalRandom.current().nextLong(), ThreadLocalRandom.current().nextLong());
```

Comment:
I don't think using `ThreadLocalRandom` to generate a 128 bit number is safe, because internally the generator only has a 64-bit seed with fixed gamma, and thus we can only get 64-bits of randomness after initialization. This means there are 2^64 unique sequences that can be produced (even if they are effectively infinite in length), with the consequence that the second value is always dependent on the first.

(I saw this happen in practice years ago when someone generated UUIDs using a similar generator with a 32-bit seed. There were a huge number of collisions.)

Also, since this needs to be secure, we should use the cryptographically secure generator, to prevent someone from predicting future slugs by reverse engineering the internal state of the generator.

We do this elsewhere as

```java
private final String slug = randomUUID().toString().replace("-", "");
```

We sometimes call lowercase, but I realize now that is not needed. I had originally been cautious as the Javadoc does not explicitly mention case. It does say

> For more information including algorithms used to create UUIDs, see RFC 4122

And RFC 4122 says

>  The hexadecimal values "a" through "f" are output as lower case characters

So it seems safe to say that the existing lowercase behavior is actually (indirectly) specified.